### ISSUE TITLE
Support binding boolean values

### ISSUE THREAD

**Issue by lukaseder** [oracle/oracle-r2dbc#31]
While Oracle SQL doesn't (yet?) support the standard SQL `BOOLEAN` type, JDBC does to some extent, mapping `BOOLEAN` values to `1=TRUE` and `0=FALSE`. Unfortunately, this is currently not possible with the oracle-r2dbc driver. This fails:

```java
System.out.println(
Flux.from(cf.create())
    .flatMap(c -> c.createStatement("select ? from dual").bind(0, true).execute())
    .flatMap(it -> it.map((r, m) -> r.get(0, Boolean.class)))
    .collectList()
    .block()
);
```

The exception being:

```
Exception in thread "main" oracle.r2dbc.impl.OracleR2dbcExceptions$OracleR2dbcException: [17004] [99999] Ungültiger Spaltentyp
	at oracle.r2dbc.impl.OracleR2dbcExceptions.toR2dbcException(OracleR2dbcExceptions.java:215)
	at oracle.r2dbc.impl.OracleR2dbcExceptions.runOrHandleSQLException(OracleR2dbcExceptions.java:247)
	at oracle.r2dbc.impl.OracleStatementImpl.setInParameters(OracleStatementImpl.java:1144)
	at oracle.r2dbc.impl.OracleStatementImpl.lambda$executeSql$2(OracleStatementImpl.java:674)
	at oracle.r2dbc.impl.OracleStatementImpl.lambda$execute$37(OracleStatementImpl.java:1342)
	at reactor.core.publisher.FluxUsingWhen.deriveFluxFromResource(FluxUsingWhen.java:119)
	at reactor.core.publisher.FluxUsingWhen.subscribe(FluxUsingWhen.java:85)
	at reactor.core.publisher.Flux.subscribe(Flux.java:8185)
	at reactor.core.publisher.FluxUsingWhen.subscribe(FluxUsingWhen.java:93)
	at reactor.core.publisher.FluxDefer.subscribe(FluxDefer.java:54)
	at reactor.core.publisher.Flux.subscribe(Flux.java:8185)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:426)
	at org.jooq.test.setup.DatabaseSetup$1.lambda$2(DatabaseSetup.java:283)
	at org.jooq.impl.Internal$1.onNext(Internal.java:433)
	at reactor.core.publisher.StrictSubscriber.onNext(StrictSubscriber.java:89)
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:120)
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79)
	at reactor.core.publisher.MonoNext$NextSubscriber.onNext(MonoNext.java:82)
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79)
	at org.reactivestreams.FlowAdapters$FlowToReactiveSubscriber.onNext(FlowAdapters.java:218)
	at oracle.jdbc.datasource.impl.OracleDataSource$ConnectionPublisher$ConnectionSubscription.emitConnection(OracleDataSource.java:2746)
	at oracle.jdbc.datasource.impl.OracleDataSource$ConnectionPublisher$ConnectionSubscription.lambda$publishConnectionAsync$0(OracleDataSource.java:2733)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:614)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:844)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
	at oracle.jdbc.driver.PhysicalConnection.lambda$createUserCodeExecutor$10(PhysicalConnection.java:11713)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:391)
	at oracle.jdbc.driver.PhysicalConnection.lambda$createUserCodeExecutor$11(PhysicalConnection.java:11711)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1176)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1647)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1614)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
	Suppressed: java.lang.Exception: #block terminated with an error
		at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:99)
		at reactor.core.publisher.Mono.block(Mono.java:1703)
		at org.jooq.testscripts.R2DBC.main(R2DBC.java:42)
Caused by: java.sql.SQLException: Ungültiger Spaltentyp
	at oracle.jdbc.driver.OraclePreparedStatement.setObjectCritical(OraclePreparedStatement.java:8722)
	at oracle.jdbc.driver.OraclePreparedStatement.setObjectInternal(OraclePreparedStatement.java:8204)
	at oracle.jdbc.driver.OraclePreparedStatement.setObject(OraclePreparedStatement.java:8173)
	at oracle.jdbc.driver.OraclePreparedStatement.setObject(OraclePreparedStatement.java:12300)
	at oracle.jdbc.driver.OraclePreparedStatement.setObject(OraclePreparedStatement.java:12249)
	at oracle.jdbc.driver.OraclePreparedStatementWrapper.setObject(OraclePreparedStatementWrapper.java:1230)
	at oracle.r2dbc.impl.OracleStatementImpl.lambda$setInParameters$29(OracleStatementImpl.java:1145)
	at oracle.r2dbc.impl.OracleR2dbcExceptions.runOrHandleSQLException(OracleR2dbcExceptions.java:244)
	... 35 more
```

Binding a `NULL` value also fails:

```java
System.out.println(
Flux.from(cf.create())
    .flatMap(c -> c.createStatement("select ? from dual").bindNull(0, Boolean.class).execute())
    .flatMap(it -> it.map((r, m) -> r.get(0, Boolean.class)))
    .collectList()
    .block()
);
```

... with a slightly different stack trace

```
Exception in thread "main" oracle.r2dbc.impl.OracleR2dbcExceptions$OracleR2dbcException: [17004] [99999] Ungültiger Spaltentyp: 16
	at oracle.r2dbc.impl.OracleR2dbcExceptions.toR2dbcException(OracleR2dbcExceptions.java:215)
	at oracle.r2dbc.impl.OracleR2dbcExceptions.runOrHandleSQLException(OracleR2dbcExceptions.java:247)
	at oracle.r2dbc.impl.OracleStatementImpl.setInParameters(OracleStatementImpl.java:1144)
	at oracle.r2dbc.impl.OracleStatementImpl.lambda$executeSql$2(OracleStatementImpl.java:674)
	at oracle.r2dbc.impl.OracleStatementImpl.lambda$execute$37(OracleStatementImpl.java:1342)
	at reactor.core.publisher.FluxUsingWhen.deriveFluxFromResource(FluxUsingWhen.java:119)
	at reactor.core.publisher.FluxUsingWhen.subscribe(FluxUsingWhen.java:85)
	at reactor.core.publisher.Flux.subscribe(Flux.java:8185)
	at reactor.core.publisher.FluxUsingWhen.subscribe(FluxUsingWhen.java:93)
	at reactor.core.publisher.FluxDefer.subscribe(FluxDefer.java:54)
	at reactor.core.publisher.Flux.subscribe(Flux.java:8185)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:426)
	at org.jooq.test.setup.DatabaseSetup$1.lambda$2(DatabaseSetup.java:283)
	at org.jooq.impl.Internal$1.onNext(Internal.java:433)
	at reactor.core.publisher.StrictSubscriber.onNext(StrictSubscriber.java:89)
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:120)
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79)
	at reactor.core.publisher.MonoNext$NextSubscriber.onNext(MonoNext.java:82)
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79)
	at org.reactivestreams.FlowAdapters$FlowToReactiveSubscriber.onNext(FlowAdapters.java:218)
	at oracle.jdbc.datasource.impl.OracleDataSource$ConnectionPublisher$ConnectionSubscription.emitConnection(OracleDataSource.java:2746)
	at oracle.jdbc.datasource.impl.OracleDataSource$ConnectionPublisher$ConnectionSubscription.lambda$publishConnectionAsync$0(OracleDataSource.java:2733)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:614)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:844)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
	at oracle.jdbc.driver.PhysicalConnection.lambda$createUserCodeExecutor$10(PhysicalConnection.java:11713)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:391)
	at oracle.jdbc.driver.PhysicalConnection.lambda$createUserCodeExecutor$11(PhysicalConnection.java:11711)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1176)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1647)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1614)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
	Suppressed: java.lang.Exception: #block terminated with an error
		at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:99)
		at reactor.core.publisher.Mono.block(Mono.java:1703)
		at org.jooq.testscripts.R2DBC.main(R2DBC.java:42)
Caused by: java.sql.SQLException: Ungültiger Spaltentyp: 16
	at oracle.jdbc.driver.OracleStatement.getInternalType(OracleStatement.java:4822)
	at oracle.jdbc.driver.OraclePreparedStatement.setNullCritical(OraclePreparedStatement.java:4564)
	at oracle.jdbc.driver.OraclePreparedStatement.setNullInternal(OraclePreparedStatement.java:4437)
	at oracle.jdbc.driver.OraclePreparedStatement.setObjectInternal(OraclePreparedStatement.java:8190)
	at oracle.jdbc.driver.OraclePreparedStatement.setObject(OraclePreparedStatement.java:8173)
	at oracle.jdbc.driver.OraclePreparedStatement.setObject(OraclePreparedStatement.java:12268)
	at oracle.jdbc.driver.OraclePreparedStatement.setObject(OraclePreparedStatement.java:12249)
	at oracle.jdbc.driver.OraclePreparedStatementWrapper.setObject(OraclePreparedStatementWrapper.java:1230)
	at oracle.r2dbc.impl.OracleStatementImpl.lambda$setInParameters$29(OracleStatementImpl.java:1145)
	at oracle.r2dbc.impl.OracleR2dbcExceptions.runOrHandleSQLException(OracleR2dbcExceptions.java:244)
	... 35 more
```

Binding an `int` works:

```java
System.out.println(
Flux.from(cf.create())
    .flatMap(c -> c.createStatement("select ? from dual").bind(0, 1).execute())
    .flatMap(it -> it.map((r, m) -> r.get(0, Boolean.class)))
    .collectList()
    .block()
);

System.out.println(
Flux.from(cf.create())
    .flatMap(c -> c.createStatement("select ? from dual").bindNull(0, Integer.class).execute())
    .flatMap(it -> it.map((r, m) -> Optional.ofNullable(r.get(0, Boolean.class))))
    .collectList()
    .block()
);
```

Producing

```
[true]
[Optional.empty]
```

Notice how I can read a `Boolean.class` from a `Row`, but I can't bind it to the `Statement`. Given that ojdbc implements `java.sql.Statement.setBoolean()`, I think this is a bug or oversight.

**Comment by Michael-A-McMahon** [comment]
Thanks for catching this.

Oracle R2DBC is inferring BOOLEAN as the SQL type for Java Boolean bind values. When Oracle R2DBC specifies a BOOLEAN type bind to Oracle JDBC, it is rejected as Oracle does not support the BOOLEAN SQL type.

To work around this in v0.2.0: 
Bind a Parameter having a SQL type that Oracle Database supports, and with a Java type for which conversion to the SQL type is supported by JDBC. The supported Java to SQL mappings for JDBC are listed in table B-5 of the JDBC 4.3 Specification.
For example, this code has the Java Boolean converted to a NUMBER:
```java
Flux.usingWhen(ConnectionFactories.get(
  "r2dbc:oracle://test:test@localhost:1521/xepdb1")
  .create(),
  connection ->
    Flux.from(connection.createStatement(
      "SELECT ? FROM sys.dual")
      .bind(0, Parameters.in(R2dbcType.NUMERIC, true))
      .execute())
    .flatMap(result ->
      result.map((row, metadata) -> row.get(0, Boolean.class))),
  Connection::close)
  .toStream()
  .forEach(System.out::println);
```

The inference to the BOOLEAN SQL type does not seem to be useful for Oracle R2DBC. If the driver is deciding what SQL type to infer, then it should choose a type that the Oracle Database supports. In fact, this is what Oracle JDBC does when PreparedStatement.setObject(int, Object) is called with a Boolean object. The Oracle JDBC driver infers the bind value’s SQL type as NUMBER. I think Oracle R2DBC should do the same.

So, I’ll plan to resolve this issue by inferring NUMER as the the SQL type for Boolean. This will allow the Statement.bind(…) method to accept Boolean values, without having to specify the NUMERIC type explicitly.

**Comment by Douglas-Surber** [comment]
What about the case where the database supports BOOLEAN, ie 23?

On Jun 30, 2021, at 12:07 PM, Michael McMahon ***@***.******@***.***>> wrote:



Thanks for catching this.

Oracle R2DBC is inferring BOOLEAN as the SQL type for Java Boolean bind values. When Oracle R2DBC specifies a BOOLEAN type bind to Oracle JDBC, it is rejected as Oracle does not support the BOOLEAN SQL type.

To work around this in v0.2.0:
Bind a Parameter having a SQL type that Oracle Database supports, and with a Java type for which conversion to the SQL type is supported by JDBC. The supported Java to SQL mappings for JDBC are listed in table B-5 of the JDBC 4.3 Specification.
For example, this code has the Java Boolean converted to a NUMBER:

Flux.usingWhen(ConnectionFactories.get(
  ***@***.***:1521/xepdb1")
  .create(),
  connection ->
    Flux.from(connection.createStatement(
      "SELECT ? FROM sys.dual")
      .bind(0, Parameters.in(R2dbcType.NUMERIC, true))
      .execute())
    .flatMap(result ->
      result.map((row, metadata) -> row.get(0, Boolean.class))),
  Connection::close)
  .toStream()
  .forEach(System.out::println);

The inference to the BOOLEAN SQL type does not seem to be useful for Oracle R2DBC. If the driver is deciding what SQL type to infer, then it should choose a type that the Oracle Database supports. In fact, this is what Oracle JDBC does when PreparedStatement.setObject(int, Object) is called with a Boolean object. The Oracle JDBC driver infers the bind value’s SQL type as NUMBER. I think Oracle R2DBC should do the same.

So, I’ll plan to resolve this issue by inferring NUMER as the the SQL type for Boolean. This will allow the Statement.bind(…) method to accept Boolean values, without having to specify the NUMERIC type explicitly.

**Comment by Michael-A-McMahon** [comment]
If BOOLEAN were supported in a future database release, then we should make sure that it supports conversions between NUMBER and BOOLEAN for bind values. If this conversion were not supported, then the new database version would be incompatible with existing versions of Oracle JDBC were NUMBER is used to represent Java Boolean bind values.

I think we would want it to work like this:
The current version of Oracle JDBC would continue to represent Boolean binds as NUMBER, and the new database would have to convert 0 to false and 1 to true.
The new version of Oracle JDBC would represent Boolean binds as BOOLEAN if the database version was one that supported the BOOLEAN type, otherwise it would continue to use the NUMBER representation for older database versions. This means the database would need to support BOOLEAN to NUMBER conversion, as existing applications may be targeting NUMBER type columns with Boolean bind values.

To be sure this clear: 
Nothing I'm writing here is a commitment from Oracle to deliver any of the functionality I've just described for the JDBC driver and database.

But we end up with both Oracle R2DBC and JDBC having to update their SQL type inference for Boolean based on the database version. I think it would be better to avoid duplicating the implementation, and have the R2DBC driver just rely on JDBC to implement the inference. This means that the R2DBC driver uses PreparedStatement.setObject(int, Object) for inferred types, and uses setObject(int, Object, SQLType) when a type is explicitly defined using io.r2ddbc.spi.Parameter.

**Comment by lukaseder** [comment]
> I think it would be better to avoid duplicating the implementation, and have the R2DBC driver just rely on JDBC to implement the inference.

Yes, I think that's usually the best approach here, to mimick or re-use the JDBC driver behaviour, now and in the future. Especially for the case where an application uses both drivers depending on the use-case. That seems to be the path of least resistance for the user, who doesn't have to think about any differences not strictly related to push/pull models when switching between the drivers.

**Comment by Michael-A-McMahon** [comment]
This is fixed by #33
